ext/session/session.c

```diff
@@ -2711,6 +2723,22 @@ static const zend_function_entry session_functions[] = {
 
 /* {{{ SessionHandlerInterface functions[]
 */
+static const zend_function_entry php_session_save_handler_iface_functions[] = {
+	PHP_ABSTRACT_ME(SessionSaveHandlerInterface, open, arginfo_session_class_open)
```

Comment:
validateId() is not needed. Your session identifier is supposed to be sufficiently random so that a collision is exceedingly unlikely and not to be expected until the end of universe.

updateTimestamp() is not needed either, just call write(). A write() handler may check itself whether the session has been changed since read, if it's worth it.

gc() is not mandatory. A cron job may do it. And in case it's not, the session handler may take care of it itself in e.g. constructor (i.e. execute `if (mt_rand(0, 10000) == 0) { … }` and delete any sessions older than a certain time). There's no need for a gc() function called from within ext/session.

And createId() … not necessary either, just grab a few pseudorandom bytes and use them.